src/main/cpp/util/photonlib2/PhotonPoseEstimator.cpp

```diff
@@ -0,0 +1,104 @@
+
+PhotonPoseEstimator::PhotonPoseEstimator(frc::AprilTagFieldLayout tags,
+                                         cv::Mat cameraMatrix, 
+                                         cv::Mat distortionCoefficients,
+                                         photonlib::PhotonCamera&& cam,
+                                         frc::Transform3d robotToCamera)
+    : m_aprilTags(tags),
+      m_cameraMatrix(cameraMatrix),
+      m_distortionCoefficients(distortionCoefficients),
+      m_camera(std::move(cam)),
+      m_robotToCamera(robotToCamera),
+      m_referencePose(frc::Pose3d()) {}
+
+std::optional<EstimatedRobotPose> PhotonPoseEstimator::Update() {
+  auto result = m_camera.GetLatestResult();
+
+  if (!result.HasTargets()) {
+    return std::nullopt;
+  }
+
+  auto targets = result.GetTargets();
+
+  // List of corners mapped from 3d space (meters) to the 2d camera screen (pixels).
+  std::vector<cv::Point3f> objectPoints;
+  std::vector<cv::Point2f> imagePoints;
+  
+  // Add all target corners to main list of corners
+  auto tagCornerToObjectPoint = [](Translation3d corner, Pose3d tagPose) {
+    Translation3d point = corner.RotateBy(tagPose.Rotation()) + tagPose.Translation();
+    return cv::Point3f(point.X().value(), point.Y().value(), point.Z().value());
+  };
+
+  for (auto target : targets) {
+    int id = target.GetFiducialId();
+    std::optional<Pose3d> tagPose = m_aprilTags.GetTagPose(id);
+    if (tagPose) {
+      std::vector<std::pair<double, double>> targetCorners = target.GetDetectedCorners();
+      for (auto& corner : targetCorners) {
+        imagePoints.emplace_back(corner.first, corner.second);
+      }
+      objectPoints.push_back(tagCornerToObjectPoint(Translation3d(-1_m, +1_m, 0_m), tagPose.value()));
+      objectPoints.push_back(tagCornerToObjectPoint(Translation3d(+1_m, +1_m, 0_m), tagPose.value()));
+      objectPoints.push_back(tagCornerToObjectPoint(Translation3d(+1_m, -1_m, 0_m), tagPose.value()));
+      objectPoints.push_back(tagCornerToObjectPoint(Translation3d(-1_m, -1_m, 0_m), tagPose.value()));
+    }
+  }
+
+  if (imagePoints.empty()) {
+    return std::nullopt;
+  }
+
+  // Use SQPnP
+  cv::Mat rvec(3, 1, cv::DataType<double>::type);
+  cv::Mat tvec(3, 1, cv::DataType<double>::type);
+
+  cv::solvePnP(objectPoints, 
+               imagePoints, 
+               m_cameraMatrix, 
+               m_distortionCoefficients, 
+               rvec, 
+               tvec, 
+               cv::SOLVEPNP_SQPNP);
+
+  return EstimatedRobotPose(Pose3d(Translation3d(meter_t{tvec.at<double>(0, 0)},
+                                                 meter_t{tvec.at<double>(1, 0)},
+                                                 meter_t{tvec.at<double>(2, 0)}), 
+                                   Rotation3d(radian_t{rvec.at<double>(0, 0)},
+                                              radian_t{rvec.at<double>(1, 0)},
+                                              radian_t{rvec.at<double>(2, 0)})), 
```

Comment:
If I recall correctly, OpenCV's rvec is a rotation vector, not roll, pitch, and yaw like Rotation3d expects.

Reply:
Looking at the docs it does use an angle-axis representation, this should be fixed in the latest commit.